flasc/visualization.py

```diff
@@ -639,12 +640,137 @@ def put_label(i):
+            axes handle for plotting. Defaults to None.
+        show_scatter (bool, optional): flag to control scatter plot.
+            Defaults to True.
+        show_bin_points (bool, optional): flag to control plot of bins.
+            Defaults to True.
+        show_confidence (bool, optional): flag to control plot of
+            confidence interval. Defaults to True.
+        alpha_scatter (float, optional): Alpha for scatter
+            plot. Defaults to 0.5.
+
+    """
+
+    # Check the length of x equals length of y
+    if len(x) != len(y):
+        raise ValueError("x and y must be the same length")
     
+    # Check that x is not empty
+    if len(x) == 0:
+        raise ValueError("x is empty")
     
 
-    
+    # Declare ax if not provided
+    if ax is None:
+        _, ax = plt.subplots()
+
+    # Put points ino dataframe
+    df = pd.DataFrame({"x": x, "y": y})
 
+    # If x_edges not provided, use 50 bins over range of x
+    if x_edges is None:
+        x_edges = np.linspace(df["x"].min(), df["x"].max(), 50)
 
+    # Define x_labels as bin centers
+    x_labels = (x_edges[1:] + x_edges[:-1]) / 2.0
 
+    # Bin data
+    df["x_bin"] = pd.cut(df["x"], x_edges, labels=x_labels)
+
+    # Get aggregate statistics
+    df_agg = df.groupby("x_bin").agg(
+        {"y": ["count", "std", "min", "max", "mean"]}
+    )
+    # Flatten column names
+    df_agg.columns = ["_".join(c) for c in df_agg.columns]
+
+    # Reset the index
+    df_agg = df_agg.reset_index()
+
+    # Add the confidence interval of the mean to df_agg
+    df_agg["y_confidence"] = 1.96 * df_agg["y_std"] / np.sqrt(df_agg["y_count"])
+
+    # Plot the mean values
+    ax.plot(df_agg.x_bin, df_agg.y_mean, color=color, label=label)
+
+    # Plot the confidence interval
+    if show_confidence:
+        ax.fill_between(
+            df_agg.x_bin,
+            df_agg.y_mean - df_agg.y_confidence,
+            df_agg.y_mean + df_agg.y_confidence,
+            color=color,
+            alpha=0.2,
+        )
+
+        # Plot a dasshed line at confidence interval
+        ax.plot(
+            df_agg.x_bin,
+            df_agg.y_mean - df_agg.y_confidence,
+            color=color,
+            alpha=0.2,
+            ls="--",
+        )
+        ax.plot(
+            df_agg.x_bin,
+            df_agg.y_mean + df_agg.y_confidence,
+            color=color,
+            alpha=0.2,
+            ls="--",
+        )
+
+    # Plot the scatter points
+    if show_scatter:
+        ax.scatter(df.x, df.y, color=color, s=10, alpha=alpha_scatter)
+
+    # Plot the bin points, scaled by the counts
+    if show_bin_points:
+        ax.scatter(
+            df_agg.x_bin,
+            df_agg.y_mean,
+            color=color,
+            s=df_agg.y_count / df_agg.y_count.max() * 20,
+            alpha=0.5,
+            marker='s'
+        )
+
+    return ax
     
+# Small little demo of data plot, can delete
+if __name__ == "__main__":
```

Comment:
This is a nice example. It would be good to apply this plot somewhere in the `flasc_cookiecutter_template` repository on actual data so that users are made aware of its existence, perhaps?

Reply:
That is a good idea, just gave myself a todo item to do that, maybe also we can use someplace in flasc's examples...

Reply:
Agree, thinking that could be a seperate issue in the cookie_cutter repo?  (To add an example?). I made a small one in the wake steering example but I think we have a broader discussion of where examples go so this could ultimately join that